src/open_clip/hf_model.py

```diff
@@ -61,17 +74,31 @@ def __init__(
         d_model = self.config.hidden_size # TODO: get d_model from config
         # different models can have different names for it
         # ?? do we use separate classes for different archs or handle it with helper funcs  
-        if (d_model == embedding_dim) and (proj is None): # do we always need a proj?
+        if (d_model == width) and (proj is None): # do we always need a proj?
             self.proj = nn.Identity()
         elif proj == 'linear':
-            self.proj == nn.Linear(d_model, embedding_dim, bias=False)
+            self.proj == nn.Linear(d_model, width, bias=False)
         elif proj == 'mlp':
             # TODO: add me
             pass
 
+        self.register_buffer('attn_mask', self.build_attention_mask(), persistent=False)
+
+    def build_attention_mask(self):
+        # lazily create causal attention mask, with full attention between the vision tokens
+        # pytorch uses additive attention mask; fill with -inf
+        mask = torch.empty(self.context_length, self.context_length)
+        mask.fill_(float("-inf"))
+        mask.triu_(1)  # zero out the lower diagonal
+        return mask
+
     def forward(self, x:TensorType) -> TensorType:
         # TODO: add attention mask?
-        out = self.transformer(input_ids=x)
+        pad = torch.full((x.shape[0], self.context_length), self.config.pad_token_id)
+        pad[:, :x.shape[1]] = x
+        x = pad
+
+        out = self.transformer(input_ids=x, attention_mask=self.attn_mask[:x.shape[0]])
```

Comment:
Don't we have x padded already in shape [bs, context_len] at this stage? Or is it a ragged tensor?
If x is padded already, one can make attention mask as `(x != self.config.pad_token_id).long()`
Otherwise it can be better to instantiate `self.tokenizer = AutoTokenizer.from_pretrained(...)` and use `self.tokenizer.pad`